core/src/test/java/fr/sncf/osrd/envelope/EnvelopeOverlayTest.java

```diff
@@ -0,0 +1,302 @@
+                        null, true,
+                        new double[]{6, 8},
+                        new double[]{6, 6}
+                ),
+                EnvelopePart.generateTimes(
+                        null, true,
+                        new double[]{8, 10},
+                        new double[]{6, 6}
+                )
+        );
+
+        var builder = EnvelopeOverlayBuilder.forward(baseEnvelope);
+        builder.startContinuousOverlay(null, true, 1);
+        assertTrue(builder.addStep(5, 5));
+        var forwardEnvelope = builder.build();
+        assertEquals(6, forwardEnvelope.size());
+        assertTrue(forwardEnvelope.spaceContinuous);
+        assertFalse(forwardEnvelope.continuous);
+        assertFalse(forwardEnvelope.physicallyAccurate);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void testLongOverlay(boolean reverse) {
+        // 4 +===+=======+=======+===+ <== base
+        //            \         /
+        //             \       / <== overlay
+        //              \     /
+        //               +---+
+        //    0  1   3   4   5   6   8
+        var baseEnvelope = Envelope.make(
+                EnvelopePart.generateTimes(
+                        null, true,
+                        new double[]{0, 1},
+                        new double[]{4, 4}
+                ),
+                EnvelopePart.generateTimes(
+                        null, true,
+                        new double[]{1, 4},
+                        new double[]{4, 4}
+                ),
+                EnvelopePart.generateTimes(
+                        null, true,
+                        new double[]{4, 6},
+                        new double[]{4, 4}
+                ),
+                EnvelopePart.generateTimes(
+                        null, true,
+                        new double[]{6, 8},
+                        new double[]{4, 4}
+                )
+        );
+
+        var builder = EnvelopeOverlayBuilder.withDirection(baseEnvelope, reverse);
+        var positions = new double[] { 3, 4, 5, 6 };
+        var speeds = new double[] { 4, 3, 3, 4 };
+        EnvelopeTestUtils.buildContinuous(builder, null, true, positions, speeds, reverse);
+        var envelope = builder.build();
+        assertEquals(4, envelope.size());
+        assertTrue(envelope.physicallyAccurate);
+    }
+
+    @Test
+    void testUnlikelyIntersection() {
+        var constSpeedEnvelope = Envelope.make(EnvelopePart.generateTimes(
+                null, true,
+                new double[]{0, 3, 4},
+                new double[]{2, 1, 0}
+        ));
+
+        var builder = EnvelopeOverlayBuilder.forward(constSpeedEnvelope);
+        builder.startContinuousOverlay(null, true, 0);
+        assertFalse(builder.addStep(1, 1));
+        assertTrue(builder.addStep(4, 1));
+        var envelope = builder.build();
+        assertEquals(2, envelope.size());
+        assertTrue(envelope.physicallyAccurate);
+    }
```

Comment:
It would be nice to add a test where the overlay tries to follow the base (like when the train tries to maintain its speed)